crates/variables/src/provider/env.rs

```diff
@@ -5,31 +5,58 @@ use async_trait::async_trait;
 
 use crate::{Key, Provider};
 
+const DEFAULT_ENV_PREFIX: &str = "SPIN_VARIABLE";
+const LEGACY_ENV_PREFIX: &str = "SPIN_CONFIG";
+
 /// A config Provider that uses environment variables.
 #[derive(Debug)]
 pub struct EnvProvider {
-    prefix: String,
+    prefix: Option<String>,
     dotenv_path: Option<PathBuf>,
     dotenv_cache: Mutex<Option<HashMap<String, String>>>,
 }
 
 impl EnvProvider {
     /// Creates a new EnvProvider.
-    pub fn new(prefix: impl Into<String>, dotenv_path: Option<PathBuf>) -> Self {
+    pub fn new(prefix: Option<impl Into<String>>, dotenv_path: Option<PathBuf>) -> Self {
         Self {
-            prefix: prefix.into(),
+            prefix: prefix.map(Into::into),
             dotenv_path,
             dotenv_cache: Default::default(),
         }
     }
 
-    fn get_sync(&self, key: &Key) -> Result<Option<String>> {
-        let env_key = format!("{}_{}", &self.prefix, key.as_ref().to_ascii_uppercase());
-        match std::env::var(&env_key) {
-            Err(std::env::VarError::NotPresent) => self.get_dotenv(&env_key),
+    fn query_env(&self, env_key: &str) -> Result<Option<String>> {
+        match std::env::var(env_key) {
+            Err(std::env::VarError::NotPresent) => self.get_dotenv(env_key),
             other => other
                 .map(Some)
-                .with_context(|| format!("failed to resolve env var {}", &env_key)),
+                .with_context(|| format!("failed to resolve env var {env_key}")),
+        }
+    }
+
+    fn get_sync(&self, key: &Key) -> Result<Option<String>> {
+        let prefix = self
+            .prefix
+            .clone()
+            .unwrap_or(DEFAULT_ENV_PREFIX.to_string());
+        let use_fallback = self.prefix.is_none();
+
+        let upper_key = key.as_ref().to_ascii_uppercase();
+        let env_key = format!("{prefix}_{upper_key}");
+
+        match self.query_env(&env_key)? {
+            None if use_fallback => {
+                let env_key = format!("{LEGACY_ENV_PREFIX}_{upper_key}");
+                match self.query_env(&env_key) {
+                    result @ Ok(Some(_)) => {
+                        eprintln!("Use of environment provided variables prefix SPIN_CONFIG is deprecated");
```

Comment:
Nit: I personally find this clearer

```suggestion
                        eprintln!("Use of deprecated prefix 'SPIN_CONFIG' for Spin variables provided by environment variables.");
```